### ISSUE TITLE
Different modes of classifier usage: "Train and deploy", "Deploy any classifier"

### ISSUE THREAD

**Issue by StephenChan** [coralnet/coralnet#150]
EDIT 2021/11: Issue details are now in [this comment](https://github.com/beijbom/coralnet/issues/150#issuecomment-958304084). I'll preserve this original post for thread consistency.

-----

Train and deploy (this is what we do now): The source maintains a classifier which is trained on the images in that source. This same classifier is deployed on any new images that are added to the source.

Deploy from other (this would be new): The source classifies its images by deploying a classifier from another source. The other source must have the same labelset.

Details:

- This mode option would be selectable in the New/Edit source page, at least for now (we might have a separate backend-control page at some point: #120 ).
- For "Deploy from other", there is of course a field to select the other source whose classifier will be used.
  - The other source must be visible to the user at the time of selection (can't select a private source that the user shouldn't be able to see).
  - The other source must be using "Train and deploy".
- The source using Deploy from other has auto-classification temporarily disabled, and shows a warning message explaining why, when any of the following situations occur:
  - The labelsets of the two sources get out of sync.
  - The other source is changed to use "Deploy from other".
- The source using Deploy from other gets its images re-annotated whenever a new classifier is trained in the 'upstream' source.
- When using Deploy from other, the upstream source sees no changes to its functionality.

**Comment by StephenChan** [comment]
Another "Deploy from other" benefit to consider:

The API gives a lot more flexibility with classification in terms of where the images live, but it doesn't let you readily set up a human-in-the-loop workflow.

CoralNet's traditional classification system does let you set up such a workflow: automatic classification results get saved as "unconfirmed" annotations and top-5 suggestions/confidence scores. These can then be manually reviewed in the annotation tool interface. You click on a point, and it shows a dropdown of those suggestions + confidence scores.

Currently, you can't get the best of both worlds. You could CSV-upload the API's classification results to a source, but those results could only be uploaded as top-1 confirmed annotations. So if you went to the annotation tool to review those results, you couldn't see the 2nd through 5th suggestions, and you couldn't see any of the confidence scores from the API.

"Deploy from other" would be a way to get the best of both worlds: flexibility in terms of where the training images vs. new images must live, and potential for a human-in-the-loop workflow.

**Comment by StephenChan** [comment]
This came up again as a key focus for NOAA and their partners. Instead of building on the posts above, I'm just going to re-specify the details in this post, but with appropriate changes - most notably, pointing to an upstream classifier ID instead of an upstream source ID. This is a draft, so it's open to suggestions.

-----

Each source must select a classification mode, with two options:

- **Train and deploy** mode (default, and this is what we currently always do): The source trains classifiers on the images in that source. The latest classifier trained in this way is deployed on any new images that are added to the source.

- **Deploy any classifier** mode (this would be new): The source classifies its images by deploying an arbitrary classifier, which *may* be from another source. It could even be a previous classifier of the same source. If another source, that source must have the same labelset.

This mode option would be a dropdown field in the 'New/Edit source' form for now (or perhaps a separate vision backend control page - part of issue #92 - if we add that in the near future). And for 'Deploy any classifier', there would be another field to specify the ID of the classifier to use (issue #206 is relevant, regarding source discoverability, but not strictly necessary yet).

Details about selecting these fields:

- If the 'deploy any' classifier lives in another source, that upstream source must be visible to the user at the time of selection (can't brute-force enter IDs to find classifiers of private sources).

- If the 'deploy any' classifier lives in another source, the labelsets of the deploying source and the upstream source must match, otherwise the form submission gets an error. For that matter, the deploying source must have a labelset defined in the first place. We could potentially offer to auto-create a matching labelset if there isn't one, but I think that's added complexity that we could consider later.

- If this form is used to select a different classifier from before, or a different classification mode from before, we should wipe all machine annotations in the source and resubmit classification for the source's unclassified images.

Consistency details:

- When the mode is 'deploy any classifier', the deploying source shouldn't be allowed to change its labelset.

- If the upstream classifier gets deleted, perhaps because the upstream labelset/extractor was changed, or because the upstream source itself was deleted, then:

  - The deploying source's machine-classification should be effectively frozen - classification calls shouldn't have any effect, and shouldn't cause server errors either.

    - The alternative is to automatically revert to 'Train and deploy' mode, but I think it's better to leave things frozen until the source admin decides to take some action.

  - The deploying source's members need to be notified somehow. Perhaps the source main / browse / backend / labelset pages can have a red-text-background notice when the upstream classifier is nonexistent. (Source members see a red background and advice on what to do; source visitors see something less showy/wordy.)

- If the upstream source changes from public to private, nothing happens. Alternatively, we could treat it as if the upstream classifier got deleted, but the 'source is visible to user' logic gets complicated because we don't know which source-admin picked the 'deploy any' classifier in the first place. Maybe it was the source admin who is also an admin of the upstream source, or maybe it was a different source admin.

- If the upstream source switches from 'Train and deploy' to 'Deploy any classifier' mode, nothing happens. The upstream source's classifiers still exist and are still available to use.

Miscellaneous:

- The source main / backend pages should display the classification mode being used, and if the mode is 'deploy any classifier', should display basic details of the classifier that's currently being used. Source name, date trained, etc.

- If the mode is 'deploy any classifier', be sure that training doesn't happen in the source, to save server resources.

- If the mode is 'deploy any classifier', the source backend page should display plots and confusion matrices for the selected classifier. However, we have to figure out how often to update the plots/confusion matrices. I believe this data is currently built using valresults files or something similar, which are currently only generated when a new classifier is trained. So we may have to rework how the plots/confusion matrices are built.

- Update the About page's Privacy Policy to clarify public/private source policies regarding 'Deploy any classifier' and the API.

**Comment by StephenChan** [comment]
So the intent of issue #537 was to extract just the minimum essentials of this issue, and then I ended up doing PR #549 which re-widened the scope to... almost exactly this issue, really (as formulated by the previous comment here). Notable caveats:

> If the upstream classifier gets deleted ... The deploying source's members need to be notified somehow.

This is done, but the notification isn't as showy as was originally proposed, which is fine for now I think.

> If the mode is 'deploy any classifier', the source backend page should display plots and confusion matrices for the selected classifier.

Yes it does, though not necessarily on the latest data. The displayed results are computed on the date the classifier was trained. To always display results from the latest re-evaluation (which may have happened following training of a subsequent classifier), we'd need to rework some things about valresults (see issue #288).

> Update the About page's Privacy Policy to clarify public/private source policies regarding 'Deploy any classifier' and the API.

Okay, so this is the one thing that was simply not done yet. I think I'll make a separate issue for the data policy in general...

Also, the original post suggested a separate backend-control page, but I just ended up adding to the Edit Source page, which feels fine for now.

Overall, I'll make sure we keep tabs on the data policy task, and I think we're good on this issue otherwise.
